[Core v1.12 beta2] Document OSI Semantic Layer artifact (#9206)
fishtown-analytics/docs.getdbt.com: website/docs/reference/artifacts/sl-manifest.md

- Deserialization: `dbt-core` and MetricFlow use different libraries for handling data serialization.
- Efficiency and performance: MetricFlow and the dbt Semantic Layer need specific semantic details from the manifest. By trimming down the information printed into `semantic_manifest.json`, the process becomes more efficient and enables faster data handling between `dbt-core` and MetricFlow.

<VersionBlock firstVersion="1.12">

> ## OSI document

Reviewer: Can we move this section after the Example section?

> 

Reviewer: To be consistent with the Semantic manifest doc, what do you think about adding something like:

```suggestion
**Produced by:** Any command that parses your project (same as the semantic manifest).
```

> Starting in <Constant name="core" /> v1.12, dbt also emits `target/osi_document.json` at parse time alongside `target/semantic_manifest.json`. The `osi_document.json` artifact is an OSI (Open Semantic Interchange) representation of your project's <Constant name="semantic_layer" />.

Reviewer: ```suggestion
Starting in <Constant name="core" /> v1.12, dbt also writes an `osi_document.json` file to your `target/` directory alongside `semantic_manifest.json` at parse time. This file represents your project's <Constant name="semantic_layer" /> in the [Open Semantic Interchange (OSI)](https://github.com/open-semantic-interchange) format, a vendor-agnostic schema for describing semantic models and metrics.
```

> 

Reviewer: There are some warnings added by the [PR](https://github.com/dbt-labs/dbt-core/pull/12955/changes#diff-a307286f8c93d8a9dbf3a863a62a6b712921d15f6b4cf9d885b15a45be68f05a) -- I think it would be useful if these warnings and their causes are documented.

> For example, run:

Reviewer: I think we can remove lines 29-40, as it's the duplicative of the description above.

The OSI document is generated by converting the full `PydanticSemanticManifest` to the OSI format. Not all dbt semantic layer constructs have an OSI equivalent, so dbt emits warnings (event code `I078`) when elements are dropped or degraded during conversion:

| Warning | Cause |
|---|---|
| `CONVERSION_METRIC_DROPPED` | Conversion metrics cannot be represented in OSI and are excluded. |
| `PRIVATE_METRIC_DROPPED` | Private metrics are not included in OSI output. |
| `NATURAL_ENTITY_DROPPED` | Natural entities have no OSI equivalent and are excluded. |
| `CUMULATIVE_SEMANTICS_LOSS` | The metric is included, but its cumulative window and grain semantics cannot be represented. |

If the converter encounters an unsupported configuration, dbt logs a `SemanticValidationFailure` warning and skips writing `osi_document.json` for that invocation.

> 

Reviewer 2: I think there are a few mental model things we need to fix here. This process is notably confusing though, sorry about that 😅 
1. `SemanticValidationFailure` get logged during validation of the semantic manifest, not the conversion of the semantic manifest to the OSI document
2. A "warning" level `SemanticValidationFailure` won't stop the semantic_manifest.json or osi_document.json from being produced. An "error" level `SemanticValidationFailure` will though
3. `MFConverterIssue` (I078) issues are logged during converting to the OSI document, however they will not block production of the `osi_document.json`

With those two things in mind, a more accurate statement would be
```suggestion
If the semantic manifest fails validation, dbt logs an `error` level `SemanticValidationFailure`  and skips writing `osi_document.json` for that invocation.
```
